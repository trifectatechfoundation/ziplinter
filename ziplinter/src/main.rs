use rc_zip_sync::ReadZip;
use std::ops::Deref;

fn main() {
    let mut args = std::env::args();
    let _ = args.next();
    let path = args.next().unwrap();

    let file = std::fs::File::open(path).unwrap();
    let archive = file.read_zip().unwrap();

    eprintln!(
        "{}",
        serde_json::to_string_pretty(&archive.deref()).unwrap()
    )
}

#[cfg(test)]
mod test {
    use super::*;

    use insta::assert_json_snapshot;
    use std::{error::Error, path::Path};

    fn process_zip_file(zip_path: &Path) -> Result<serde_json::Value, Box<dyn Error>> {
        let file = std::fs::File::open(zip_path).unwrap();
        let archive = file.read_zip();

        Ok(serde_json::to_value(archive?.deref())?)
    }

    #[test]
    fn snapshot_zip_files() {
        let fixtures_dir = std::env::current_dir()
            .unwrap()
            .join("../testdata")
            .canonicalize()
            .unwrap();

        println!("fixtures_dir: {}", fixtures_dir.display());

        for entry in std::fs::read_dir(fixtures_dir).expect("Failed to read fixtures directory") {
            let entry = entry.expect("Failed to read entry");
            let path = entry.path();

            if path
                .extension()
                .is_some_and(|e| e.eq_ignore_ascii_case("zip"))
            {
                let Ok(result) = process_zip_file(&path) else {
                    continue;
                };
                println!(
                    "current file: {}",
                    Path::new(path.file_name().unwrap()).display()
                );
                assert_json_snapshot!(
                    format!("{}", Path::new(path.file_name().unwrap()).display()),
                    result
                );
            }
        }
    }
}
