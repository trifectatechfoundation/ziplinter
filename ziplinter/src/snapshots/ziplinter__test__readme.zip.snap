---
source: ziplinter/src/main.rs
expression: result
---
{
  "comment": "This is the source code repository for the Go programming language.  \r\n\r\nFor documentation about how to install and use Go,\r\nvisit https://golang.org/ or load doc/install.html in your web browser.\r\n\r\nAfter installing Go, you can view a nicely formatted\r\ndoc/install.html by running godoc --http=:6060\r\nand then visiting http://localhost:6060/doc/install.html.\r\n\r\nUnless otherwise noted, the Go source files are distributed\r\nunder the BSD-style license found in the LICENSE file.\r\n\r\n--\r\n\r\nBinary Distribution Notes\r\n\r\nIf you have just untarred a binary Go distribution, you need to set\r\nthe environment variable $GOROOT to the full path of the go\r\ndirectory (the one containing this README).  You can omit the\r\nvariable if you unpack it into /usr/local/go, or if you rebuild\r\nfrom sources by running all.bash (see doc/install.html).\r\nYou should also add the Go binary directory $GOROOT/bin\r\nto your shell's path.\r\n\r\nFor example, if you extracted the tar file into $HOME/go, you might\r\nput the following in your .profile:\r\n\r\n    export GOROOT=$HOME/go\r\n    export PATH=$PATH:$GOROOT/bin\r\n\r\nSee doc/install.html for more details",
  "contents": [
    {
      "central": {
        "comment": "",
        "compressed_size": 598,
        "crc32": 2181565152,
        "creator_version": {
          "host_system": "Unix",
          "version": 30
        },
        "disk_nbr_start": 0,
        "external_attrs": 2175008768,
        "extra": [
          85,
          84,
          5,
          0,
          3,
          75,
          201,
          127,
          76,
          117,
          120,
          11,
          0,
          1,
          4,
          232,
          3,
          0,
          0,
          4,
          232,
          3,
          0,
          0
        ],
        "flags": 0,
        "header_offset": 0,
        "internal_attrs": 1,
        "method": "Deflate",
        "mode": 420,
        "modified": "2010-09-02T15:56:59Z",
        "name": "README",
        "reader_version": {
          "host_system": "MsDos",
          "version": 20
        },
        "uncompressed_size": 1096
      },
      "local": {
        "accessed": null,
        "compressed_size": 598,
        "crc32": 2181565152,
        "created": null,
        "extra": [
          85,
          84,
          9,
          0,
          3,
          75,
          201,
          127,
          76,
          146,
          245,
          161,
          76,
          117,
          120,
          11,
          0,
          1,
          4,
          232,
          3,
          0,
          0,
          4,
          232,
          3,
          0,
          0
        ],
        "flags": 0,
        "gid": 1000,
        "header_offset": 0,
        "method": "Deflate",
        "method_specific": "None",
        "mode": 0,
        "modified": "2010-09-02T15:56:59Z",
        "name": "README",
        "reader_version": {
          "host_system": "MsDos",
          "version": 20
        },
        "uid": 1000,
        "uncompressed_size": 1096
      }
    }
  ],
  "encoding": "Utf8",
  "eocd": {
    "dir": {
      "inner": {
        "dir_disk_nbr": 0,
        "dir_records_this_disk": 1,
        "directory_offset": 662,
        "directory_records": 1,
        "directory_size": 76,
        "disk_nbr": 0
      },
      "offset": 738
    },
    "dir64": null,
    "global_offset": 0
  },
  "parsed_ranges": [
    {
      "contains": "end of central directory record",
      "end": 1885,
      "start": 738
    },
    {
      "contains": "central directory header",
      "end": 738,
      "filename": "README",
      "start": 662
    },
    {
      "contains": "local file header",
      "end": 64,
      "filename": "README",
      "start": 0
    },
    {
      "contains": "file data",
      "end": 662,
      "filename": "README",
      "start": 64
    }
  ],
  "size": 1886
}
